File: src/SnoDb/AddOrReplaceSaveAction.vb
Imports System.IO
Imports System.IO.Compression
Imports Newtonsoft.Json

Public Class AddOrReplaceSaveAction
    Implements ISaveAction

    Public ReadOnly Property ItemPath As String
    Public ReadOnly Property ItemJson As String

    Public Sub New(itemPath As String, item As Object)
        Me.ItemPath = itemPath
        Me.ItemJson = JsonConvert.SerializeObject(item)
    End Sub

    Public Async Function SaveAsync(archive As ZipArchive) As Task Implements ISaveAction.SaveAsync
        Dim entry = archive.GetEntry(Me.ItemPath)
        If entry IsNot Nothing Then
            entry.Delete()
        End If

        entry = archive.CreateEntry(Me.ItemPath)
        Using writer = New StreamWriter(entry.Open())
            Await writer.WriteAsync(Me.ItemJson)
        End Using
    End Function

End Class

File: src/SnoDb/AddOrReplaceSaveAction.cs
using System.IO;
using Newtonsoft.Json;
using Ionic.Zip;

namespace Symmex.SnoDb
{
    /// <summary>
    /// Handles operations to add or replace a new entry in the zip archive  
    /// </summary>
    public class AddOrReplaceSaveAction : ISaveAction
    {
        public string ItemPath { get; }
        public string ItemJson { get; }

        public AddOrReplaceSaveAction(string path, object item)
        {
            ItemPath = path;
            ItemJson = JsonConvert.SerializeObject(item, SnoDbConfig.DefaultSerializerSettings);
        }

        public void Save(ZipFile archive)
        {
            if(archive.ContainsEntry(ItemPath))
                archive.RemoveEntry(ItemPath);

            var entry =  archive.AddEntry(ItemPath, ItemJson);
        }
    }
}
